tensorflow_probability/python/math/special.py

```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
 
 
 def _betainc_naive(a, b, x):
-  """Returns the regularized incomplete beta function elementwise."""
+  """Returns the regularized incomplete beta function element-wise."""
   dtype = dtype_util.common_dtype([a, b, x], tf.float32)
   a = tf.convert_to_tensor(a, dtype=dtype)
   b = tf.convert_to_tensor(b, dtype=dtype)
   x = tf.convert_to_tensor(x, dtype=dtype)
-  broadcast_shape = ps.broadcast_shape(
-      ps.shape(a), ps.shape(b))
-  broadcast_shape = ps.broadcast_shape(
-      broadcast_shape, ps.shape(x))
+
+  broadcast_shape = functools.reduce(
+      ps.broadcast_shape, [ps.shape(a), ps.shape(b), ps.shape(x)])
+
   a = tf.broadcast_to(a, broadcast_shape)
   b = tf.broadcast_to(b, broadcast_shape)
   x = tf.broadcast_to(x, broadcast_shape)
-  return tf.math.betainc(a, b, x)
 
-# Derivative implementation based on
-# [1] R. Boik, J. Robinson-Cox,
-#     Derivatives of the Incomplete Beta Function
-#     https://www.jstatsoft.org/article/view/v003i01/beta.der.pdf
+  return tf.math.betainc(a, b, x)
 
 
-def partial_numerator(a, b, x, i):
-  """Partial numerator used in continued fraction expansion of betainc."""
-  f = b * x / (a * (1 - x))
-  result = (a + b + i - 2.) * (a + i - 1) * (b - i) / (
-      (a + 2 * i - 3.) * tf.math.square(a + 2 * i - 2.) * (a + 2 * i - 1.))
-  result = result * tf.math.square(a * f / b) * (i - 1.)
-  return tf.where(
-      tf.math.equal(i, 1.),
-      a * f * (b - 1.) / (b * (a + 1.)),
-      result)
+def _betainc_nth_partial_numerator(iteration, a, b, x, dtype):
+  """Partial numerator used in the continued fraction for betainc."""
+  zero = tf.constant(0., dtype=dtype)
+  one = tf.constant(1., dtype=dtype)
+  two = tf.constant(2., dtype=dtype)
+
+  iteration_is_even = tf.equal(iteration % two, zero)
+
+  # The nth partial numerator is given here: https://dlmf.nist.gov/8.17#E23
+  def even_partial_numerator():
+    # d_{2m}
+    n = iteration / two
+    a_plus_2n = a + two * n
+    a_plus_2n_minus_one = a_plus_2n - one
+    denominator = a_plus_2n * a_plus_2n_minus_one
+
+    db = n * x / denominator
+    value =  db * (b - n)
+    da = -value * (a_plus_2n + a_plus_2n_minus_one) / denominator
+    gradient = tf.concat([da, db], axis=-1)
+    return value, gradient
+
+  def odd_partial_numerator():
+    # d_{2m+1}
+    n = (iteration - one) / two
+    a_plus_n = a + n
+    a_plus_2n = a_plus_n + n
+    a_plus_2n_plus_one = a_plus_2n + one
+    a_plus_b_plus_n = a_plus_n + b
+    denominator = a_plus_2n * a_plus_2n_plus_one
+
+    db = -a_plus_n * x / denominator
+    value =  db * a_plus_b_plus_n
+    da = -value * ((a_plus_2n + a_plus_2n_plus_one) / denominator) - x * (
+        two * a_plus_n + b) / denominator
+    gradient = tf.concat([da, db], axis=-1)
+    return value, gradient
+
+  # Assume iteration is always greater than 1.
+  return tf.cond(
```

Comment:
If we use `tf.where` instead of `tf.cond`, I understand that the tuple `(values, gradients)` will be computed twice for each call of `_betainc_nth_partial_numerator`, regardless of whether `iteration` is odd or even. Is my understanding correct? If so, why should we prefer `tf.where` over `tf.cond`?

Reply:
In general we prefer using `tf.where`, especially in cases like this that are relatively cheap. `tf.cond` creates a runtime branch to decide which expression to compute. This can be expensive, and often `tf.where` (which is branchless) will be more performant on accelators like GPU or Cloud TPU.

Reply:
After refactoring the continued fraction evaluation procedures, `tf.cond` is no longer needed.